### DISCUSSION TITLE
Protecting .github/workflows

### DISCUSSION THREAD

**Discussion by Fabien-Chouteau** [community/community#25236] [Actions]
<p>Hi all,</p>
<p>It is mentioned on<a href="https://developer.github.com/actions/managing-workflows/" rel="nofollow noopener">this page</a> that it is possible to restrict modifications of the workflow file.</p>
<p>However I there is no explanation on how to do it. Is there a doc somewhere?</p>
<p>I dig the actions/workflow framework but I am a little worried that anyone can change a workflow with a pull-request and expose secrets or disable checks.</p>
<p>Thanks in advance,</p>

**Comment by jeremyepling** [comment, marked as answer]
<p>Those are the old docs for workflows based on HCL. You should look at the docs for the new YAML workflows because a lot has changed. <a href="https://help.github.com/en/categories/automating-your-workflow-with-github-actions" rel="nofollow noopener">https://help.github.com/en/categories/automating-your-workflow-with-github-actions</a></p>
<p>There are a few ways to control changes to workflows flows:</p>
<ul>
<li>Workflow files will only execute in the context of a branch they exist in. If someone creates a PR with a change for a workflow file that targets <code>master</code>, it won’t run on master until that PR is merged into <code>master</code>.</li>
<li>You can use <a href="https://help.github.com/en/articles/about-code-owners" rel="nofollow noopener">codeowners</a> and branch protection to require reviews for changes to files in .github/workflows</li>
<li>Secrets aren’t passed to workflows that run in PR from forks, to prevent leaks.</li>
</ul>

**Comment by Fabien-Chouteau** [comment]
<p>Thank you <a class="mention" href="/jeremyepling">@jeremyepling</a> </p>
<p>I will play with codeowners and see if works for me.</p>

**Comment by gabriel-kohen-by** [comment]
<p>Hi,</p>
<p>Does it mean that if you have an org that folks can contribute to private repository by creating branches and waiting for it to be reviewed before merging the PR to master, can go ahead create a branch, change the a workflow on the branch (allowing them to read secrets), and if the workflow changes live deployments there is no protection we can put.</p>
<p>That eliminates the whole development workflow of having contirbutors helping collaborators on a repo and allowing them to make changes without a PR being approved and merged to master.</p>
<p>Any way of protecting these?</p>

**Comment by gabriel-kohen-by** [comment]
<p>OK. I’ve ran few tests. It seems like you first have to make sure you have a workflow that is looking at non-master (or whatever protected branches you have) to be able to run on them.</p>
<p>You cannot create a branch, add a workflow in that branch and have it trigger on that branch if it’s not already merged to master(or default) which is great.</p>
<p>You just have to make sure you don’t accidently introduce workflow that can change your production env on non master branches(or production branches) -duh :slight_smile:</p>

**Comment by kceb** [comment]
<p>Let’s say we have a deployment workflow that uses a secret key stored in github secrets. Can’t a dev with write-access make a branch off master, create a new workflow that runs on push to their custom branch, utilize the secret key that gets passed in, and trigger a deploy? How do you protect against that?</p>
<p>Ideally I would like a dev to have write permissions but not deploy permissions</p>
<p>I’d like to avoid having all internal devs fork the repo</p>
<p>Edit: upon further investigation it seems <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories" rel="nofollow noopener">like forks won’t run any workflows</a>. That is a blocker for us because we need tests/linting to run on PRs</p>

**Comment by hu13** [comment]
<p><code>You cannot create a branch, add a workflow in that branch and have it trigger on that branch if it's not already merged to master(or default) which is great.</code> </p>
<p><a class="mention" href="/gabriel-kohen-jdas">@gabriel-kohen-jdas</a> </p>
<p>How would you achieve the above which you said earlier?</p>
<p>Because I am able to do the exact opposite.</p>

**Comment by shinebayar-g** [comment]
<p>I don’t know how to deal with this situation.</p>
<p>Let’s say I’ve a single workflow setup that runs on push to master only. But someone just forks the repo, add their own modified version of workflow and creates PR against master with <code>on.pull\_request</code> policy. Even tho secrets are not exposed to that PR, it’s still annoying if that PR contains something bad. Also even if base repo workflow contains policy for <code>on.pull_request</code> event, forked PR can still override base repo’s workflows. </p>
<p>Am I missing something here? I’m not sure this is the desired behaviour.</p>

**Comment by gtirloni** [comment]
<p>I’m on the same boat. I have a public repository that anyone can fork and submitted PRs to and right now anyone can change the .github/workflow files as they please and it’ll get executed just fine. With GitHub-owned runners that is less of a concern but with self-hosted runners this is a major security flaw.</p>
<p>In Jenkins, we have a list of pre-approved users that will get their PRs built automatically and anyone else needs an okay from an admin before that happens. I don’t know how to implement this safely in GitHub Actions.</p>
<p>I played with some actions that limit builds to a certain user list but, because the user list is inside the workflow file, contributors can simply erase that from the file and be fine. It’s unsettling.</p>
<p>I think we need a flag somewhere saying the only workflows that will run are the ones in the master branch and not ones in other branches and/or PR’s.</p>

**Comment by bittelc** [comment]
<p>As is already mentioned in prior comments, this question should not be marked “Solved”.</p>
<p>An example of the problem:<br>
I have enabled GitHub Actions to deploy to my production environment. To do this, GitHub Actions checks for:</p>
<pre><code class="lang-auto">if: github.ref != 'refs/heads/master'

</code></pre>
<p>However, there is no way to prevent a malicious PR to my repository which simply removes this line, thereby accomplishing a production deployment without the PR being merged.</p>

**Comment by MalloZup** [comment]
<p>I think and agree with <a class="mention" href="/bittelc">@bittelc</a> </p>
<p>So the issue is not solved.</p>
<p>And this issue should be differianted in 2 category.</p>
<ul>
<li>
<p>workflow on branches (yes this can be fixed easy)</p>
</li>
<li>
<p>workflow on Pullrequests  (this is not fixed)</p>
</li>
</ul>
<p>The solution here proposed is only for workflow running on Branches.</p>
<p>However if you have a workflow running for pull-request, there is no way by github to proctect a user to change the workflow for PR and redirect output.</p>

**Comment by ghost** [comment]
<p><a class="mention" href="/jeremyepling">@jeremyepling</a> this issue is very much not resolved today. Here is one example of how the protection around the .github/workflow files are not sufficient. As <a class="mention" href="/mallozup">@mallozup</a> said:</p>
<p>If a project has a workflow (example.yml) that is triggered on push to master, and I create a branch in the project and change the workflow example.yml to be triggered on pull_request against master and then change the workflow itself I can circumvent all branch protection and controls on the pipeline and use it to execute whatever I want and utilize the projects secrets.</p>

**Comment by MalloZup** [comment]
<p>I have opened an issue to runner repo regarding this:</p>
<aside class="onebox githubissue">
  <header class="source">
      <a href="https://github.com/actions/runner/issues/494" target="_blank" rel="nofollow noopener">github.com/actions/runner</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/runner/issues/494" target="_blank" rel="nofollow noopener">Add functionality for allow self-hosted runner to protect workflow file and allow only execution only for collaborators for PRS</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-05-20" data-time="16:41:26" data-timezone="UTC">04:41PM - 20 May 20 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/MalloZup" target="_blank" rel="nofollow noopener">
          <img alt="MalloZup" src="https://user-images.githubusercontent.com/10886597/181089776-7aa7c28f-db57-4b65-97e5-1e6033d357c9.png" class="onebox-avatar-inline" width="20" height="20">
          MalloZup
        </a>
      </div>
    </div>
  </div>
</div>

<div class="github-row">
  <p class="github-content">Description:
Hi all,
first of all I really like the github action runner and self-hosted runner.
It has however an issue with working for...</p>
</div>

<div class="labels">
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">enhancement</span>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by eladchen** [comment]
<p>Hi Everyone,</p>
<p>I’ve put together a Github App that will trigger a workflow based on the changed files, and the actor behind the change.</p>
<p>I’m looking for a few early adopters to provide feedback before releasing it.<br>
If you are interested please let me know, and I’ll setup a place for us to talk.</p>

**Comment by eladchen** [comment]
<p>The app has been released.</p>
<p><a href="https://medium.com/@eladchen/protecting-your-github-workflows-b4359683bd69?sk=d59c21f5edc62ffcd5e0d3b14bbedd35" class="onebox" target="_blank" rel="noopener nofollow ugc">https://medium.com/@eladchen/protecting-your-github-workflows-b4359683bd69?sk=d59c21f5edc62ffcd5e0d3b14bbedd35</a></p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/apps/protected-workflows" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/3628913/181089778-684cf7a5-1196-409f-8610-6b06476c2430.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://github.com/apps/protected-workflows" target="_blank" rel="noopener nofollow ugc">Build software better, together</a></h3>

<p>GitHub is where people build software. More than 50 million people use GitHub to discover, fork, and contribute to over 100 million projects.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by PeteMac88** [comment]
<p>Hey there,<br>
any news here because the missing security features is causing some workarounds for us in our project. Does everyone has a good way of prevent specific roles or users to edit the workflow file and so get access to the repo secrets?</p>
<p>Best regards Peter</p>

**Comment by worthy7** [comment]
<p>Hi, so what is the conclusion. I just set up Github actions to deploy to firebase<br>
<a href="https://firebase.google.com/docs/hosting/github-integration" class="onebox" target="_blank" rel="noopener nofollow ugc">https://firebase.google.com/docs/hosting/github-integration</a></p>
<p>But I also noticed that anyone could just edit the PR action to deploy to production instead and make a PR …</p>

**Comment by worthy7** [comment]
<p>Perhaps PR/test and master/prod should be handled differently.</p>
<p>In the PR case, a new PR and use Github actions to spin up a new server and deploy it, using a service account that does NOT have access to production stuff.</p>
<p>Then when merging into master, instead, the production server can watch for webhooks and then pull/build/deploy it.</p>
<p>This would solve the issue.</p>

**Comment by RHV044** [comment]
<p>Hey everyone, have been any updates on this or is this still an issue?</p>

**Comment by kroussou** [comment]
<p>Possible workaround is to use <a href="https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/using-openid-connect-with-reusable-workflows" rel="noopener nofollow ugc">OpenID connect with reusable workflows</a> and external secret storage like Azure KeyVault. If you deploy directly to the cloud you may need no secrets at all.</p>
<p>Once you add check for valid <code>job_workflow_ref</code> as part of <a href="https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#defining-trust-conditions-on-cloud-roles-using-oidc-claims" rel="noopener nofollow ugc">trust conditions</a> between GitHub and cloud provider, only trusted workflows will be allowed to access secrets or other cloud resources.</p>
<p>You can’t prevent workflow modification this way but modified workflows would not have access to sensitive data.</p>

**Comment by 13013SwagR** [comment]
I opened a feature request that would solve this and seems fairly simple to implement, see https://github.com/orgs/community/discussions/53430
